### PR TITLE
Added overload for deadletterAsync to specify dead letter reason

### LINKED ISSUE #339: Add user properties when deadletter
In the old service bus API, you could specify a reason and error description when moving a message to the DLQ:
`public void DeadLetter (string deadLetterReason, string deadLetterErrorDescription);`

## Issue
I am trying to replicate that behaviour with this new service bus API:

        async Task ProcessMessagesAsync(Message message, CancellationToken token)
        {
            try
            {
                //Attempt to process message.....
                ProcessMessage(message);

                //Message processed successfully
                await _subscriptionClient.CompleteAsync(message.SystemProperties.LockToken);
            }
            catch (Exception ex)
            {
                //Error occurred during processing, note the error and move to DLQ
                message.UserProperties.Add("ErrorMessage", ex.ToString());
                await _subscriptionClient.DeadLetterAsync(message.SystemProperties.LockToken);
            }
        }

However, while the UserProperties.Add method succeeds, they are not actually persisted when the message is moved to the DLQ.

## Question
Is this a bug, or is this expected behaviour and there is a different way to achieve the same thing?

### PR DESCRIPTION
**Pull request by nemakam**
Added overload for deadletterAsync to specify dead letter reason.
Resolves #339

### CODE REVIEW

**Comment by SeanFeldman** [comment]
@nemakam wow, you did everything 🙂
Will look at this PR, thank you.

File: test/Microsoft.Azure.ServiceBus.UnitTests/SenderReceiverTests.cs [hunk 1]
```diff
@@ -1,6 +1,9 @@
 // Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using Microsoft.Azure.Amqp;
+using Microsoft.Azure.ServiceBus.Amqp;
```
Comment (vinaysurya) [inline test/Microsoft.Azure.ServiceBus.UnitTests/SenderReceiverTests.cs]: nit: move inside

File: src/Microsoft.Azure.ServiceBus/Message.cs [hunk 2]
```diff
@@ -13,6 +13,16 @@ namespace Microsoft.Azure.ServiceBus
     /// </summary>
     public class Message
     {
+        /// <summary>
+        /// User property key representing dead letter reason, when a message is received from a dead-letter subqueue of an entity.
```
Comment (vinaysurya) [inline src/Microsoft.Azure.ServiceBus/Message.cs]: >dead letter [](start = 43, length = 11)

nit: deadletter should be one word

Comment (vinaysurya) [inline src/Microsoft.Azure.ServiceBus/Message.cs]: >dead-letter [](start = 97, length = 11)

nit: same here..one word

File: src/Microsoft.Azure.ServiceBus/Core/IReceiverClient.cs [hunk 3]
```diff
@@ -101,5 +102,21 @@ public interface IReceiverClient : IClientEntity
...
+        /// <param name="deadLetterErrorDescription">The error description for deadlettering the message.</param>
+        /// <param name="propertiesToModify">The properties of the message to modify while moving to sub-queue.</param>
+        /// <remarks>
+        /// A lock token can be found in <see cref="Message.SystemPropertiesCollection.LockToken"/>,
+        /// only when <see cref="ReceiveMode"/> is set to <see cref="ServiceBus.ReceiveMode.PeekLock"/>.
+        /// In order to receive a message from the deadletter queue, you will need a new <see cref="IMessageReceiver"/>, with the corresponding path.
+        /// You can use <see cref="EntityNameHelper.FormatDeadLetterPath(string)"/> to help with this.
+        /// This operation can only be performed on messages that were received by this receiver.
+        /// </remarks>
+        Task DeadLetterAsync(string lockToken, string deadLetterReason, string deadLetterErrorDescription = null, IDictionary<string, object> propertiesToModify = null);
```
Comment (vinaysurya) [inline src/Microsoft.Azure.ServiceBus/Core/IReceiverClient.cs]: >DeadLetterAsync [](start = 13, length = 15)

Can we please check if Abandon/Defer have similar reason and properties API surface to be exposed

File: src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs [hunk 4]
```diff
@@ -966,39 +1002,60 @@ protected virtual Task OnCompleteAsync(IEnumerable<string> lockTokens)
...
+            if (deadLetterReason != null && propertiesToModify != null && propertiesToModify.ContainsKey(Message.DeadLetterReasonHeader))
+            {
+                throw new ArgumentException(nameof(propertiesToModify), $"Cannot provide {nameof(deadLetterReason)} param and modify property {Message.DeadLetterReasonHeader} at the same time");
+            }
+
+            if (deadLetterErrorDescription != null && propertiesToModify != null && propertiesToModify.ContainsKey(Message.DeadLetterErrorDescriptionHeader))
+            {
+                throw new ArgumentException(nameof(propertiesToModify), $"Cannot provide {nameof(deadLetterErrorDescription)} param and modify property {Message.DeadLetterErrorDescriptionHeader} at the same time");
+            }
+
```
Comment (vinaysurya) [inline src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs]: now that its either this or that API, do you think we still need these checks?

**Comment by nemakam** [comment]
@SeanFeldman 
> Will look at this PR, thank you.

Sorry for closing the PR soon. If you have any comments, feel free to leave them. I'll send another PR if required.

**Comment by sfalik** [comment]
Fantastic, thanks so much!

File: src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs [hunk 5]
```diff
@@ -966,39 +1002,50 @@ protected virtual Task OnCompleteAsync(IEnumerable<string> lockTokens)
...
-            return this.DisposeMessagesAsync(lockTokens, new Modified { UndeliverableHere = true });
+            return this.DisposeMessagesAsync(lockTokens, GetDeferOutcome(propertiesToModify));
         }
 
-        protected virtual Task OnDeadLetterAsync(string lockToken)
+        protected virtual Task OnDeadLetterAsync(string lockToken, IDictionary<string, object> propertiesToModify = null, string deadLetterReason = null, string deadLetterErrorDescription = null)
         {
+            if (deadLetterReason != null && deadLetterReason.Length > Constants.MaxDeadLetterReasonLength)
+            {
+                throw new ArgumentOutOfRangeException(nameof(deadLetterReason), $"Max permitted length is {Constants.MaxDeadLetterReasonLength}");
```
Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs]: >Max
Maximum?

File: src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs [hunk 6]
```diff
@@ -966,39 +1002,50 @@ protected virtual Task OnCompleteAsync(IEnumerable<string> lockTokens)
...
+            }
+
+            if (deadLetterErrorDescription != null && deadLetterErrorDescription.Length > Constants.MaxDeadLetterReasonLength)
+            {
+                throw new ArgumentOutOfRangeException(nameof(deadLetterErrorDescription), $"Max permitted length is {Constants.MaxDeadLetterReasonLength}");
```
Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs]: same here

File: src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs [hunk 7]
```diff
@@ -1194,6 +1241,39 @@ static void CloseRequestResponseSession(RequestResponseAmqpLink requestResponseA
...
+                    foreach (var pair in propertiesToModify)
+                    {
+                        object amqpObject;
+                        if (AmqpMessageConverter.TryGetAmqpObjectFromNetObject(pair.Value, MappingType.ApplicationProperty, out amqpObject))
+                        {
+                            amqpPropertiesToModify[new MapKey(pair.Key)] = amqpObject;
+                        }
+                    }
+
+                    if (amqpPropertiesToModify.Count > 0)
```
Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs]: Looking at the [logic above](https://github.com/Azure/azure-service-bus-dotnet/pull/341/files?diff=unified#diff-f0c668c8659c9b5c15eb016bcd714028R1260) there's a chance that not all properties could end up in the `amqpPropertiesToModify` collection. In that case, should there be a check and a warning logged if it happens?

File: src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs [hunk 8]
```diff
@@ -1194,6 +1241,39 @@ static void CloseRequestResponseSession(RequestResponseAmqpLink requestResponseA
...
+                {
+                    amqpRequestMessage.Map[ManagementConstants.Properties.DeadLetterDescription] = deadLetterDescription;
+                }
+
+                if (propertiesToModify != null)
+                {
+                    var amqpPropertiesToModify = new AmqpMap();
+                    foreach (var pair in propertiesToModify)
+                    {
+                        object amqpObject;
```
Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs]: `amqpObject` can be inlined

File: src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs [hunk 9]
```diff
@@ -1316,5 +1396,71 @@ void ThrowIfSessionLockLost()
...
+                if (deadLetterErrorDescription != null)
+                {
+                    rejected.Error.Info.Add(Message.DeadLetterErrorDescriptionHeader, deadLetterErrorDescription);
+                }
+
+                if (propertiesToModify != null)
+                {
+                    foreach (var pair in propertiesToModify)
+                    {
+                        object amqpObject;
```
Comment (SeanFeldman) [inline src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs]: can be inlined

**Pull request merged**